baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LambdaMethodReference.java

```diff
@@ -91,49 +89,97 @@ public Description matchLambdaExpression(LambdaExpressionTree tree, VisitorState
+                && (root.getParameters().size() == 1 || root.getParameters().isEmpty())) {
+            return convertSuppliersAndVariableInstanceMethods(methodSymbol, methodInvocation, root, state);
+        }
+
+        if (methodInvocation.getArguments().size() == root.getParameters().size()) {
+            return convertMethodInvocations(methodSymbol, methodInvocation, root, state);
+        }
+
+        return Description.NO_MATCH;
+    }
+
+    private Description convertSuppliersAndVariableInstanceMethods(
+            Symbol.MethodSymbol methodSymbol,
+            MethodInvocationTree methodInvocation,
+            LambdaExpressionTree root,
+            VisitorState state) {
+        if (!root.getParameters().isEmpty()) {
+            Symbol paramSymbol = ASTHelpers.getSymbol(Iterables.getOnlyElement(root.getParameters()));
+            Symbol receiverSymbol = ASTHelpers.getSymbol(ASTHelpers.getReceiver(methodInvocation));
+            if (!paramSymbol.equals(receiverSymbol)) {
+                return Description.NO_MATCH;
+            }
+        }
+
+        ExpressionTree receiver = ASTHelpers.getReceiver(methodInvocation);
+        boolean isLocal = receiver == null;
+        if (!isLocal && !(receiver instanceof IdentifierTree)) {
             return Description.NO_MATCH;
         }
 
         return buildDescription(root)
                 .setMessage(MESSAGE)
-                .addFix(buildFix(methodSymbol, methodInvocation, root, state))
+                .addFix(buildFix(methodSymbol, methodInvocation, root, state, isLocal))
                 .build();
     }
 
-    private static boolean shouldIgnore(
-            Symbol.MethodSymbol methodSymbol, LambdaExpressionTree root, MethodInvocationTree methodInvocation) {
-        if (!methodSymbol.isStatic()) {
-            if (root.getParameters().size() == 1) {
-                Symbol paramSymbol = ASTHelpers.getSymbol(Iterables.getOnlyElement(root.getParameters()));
-                Symbol receiverSymbol = ASTHelpers.getSymbol(ASTHelpers.getReceiver(methodInvocation));
-                return !paramSymbol.equals(receiverSymbol);
-            }
-            return true;
+    private Description convertMethodInvocations(
+            Symbol.MethodSymbol methodSymbol,
+            MethodInvocationTree methodInvocation,
+            LambdaExpressionTree root,
+            VisitorState state) {
+        List<Symbol> methodParams = getSymbols(methodInvocation.getArguments());
+        List<Symbol> lambdaParam = getSymbols(root.getParameters());
+
+        // We are guaranteed that all of root params are symbols so equality should handle cases where methodInvocation
+        // arguments are not symbols or are out of order
+        if (!methodParams.equals(lambdaParam)) {
+            return Description.NO_MATCH;
         }
-        return !root.getParameters().isEmpty();
+
+        boolean isLocal = ASTHelpers.getReceiver(methodInvocation) == null;
```

Comment:
I think we should handle a receiver IdentifierTree with name “this” as well. Worth a test case if we don’t already have one (review on thumbs)